src/subset.rs

```diff
@@ -322,6 +332,26 @@ impl<N: NodeIdT + Rand> Subset<N> {
         Ok(step)
     }
 
+    /// Performs a destructive 0-copy read of the current broadcast result value of a given proposer.
+    fn take_broadcast_result(&mut self, proposer_id: &N) -> Option<ProposedValue> {
+        let mut remove_insertion = false;
+        let value = match self
+            .broadcast_results
+            .insert(proposer_id.clone(), BroadcastResult::Taken)
+        {
+            Some(BroadcastResult::Value(value)) => Some(value),
+            Some(BroadcastResult::Taken) => None,
+            None => {
+                remove_insertion = true;
+                None
+            }
+        };
+        if remove_insertion {
+            self.broadcast_results.remove(proposer_id);
+        }
+        value
```

Comment:
If you add a `BroadcastResult::take` method (that replaces the instance with `Taken` and returns the `Option<ProposedValue>`), this method could just be

```rust
self.broadcast_results.get_mut(proposer_id).and_then(BroadcastResult::take)
```

without cloning the ID, and without inserting and removing again.

Reply:
…but then we might as well just use `Option` again, instead of `BroadcastResult`, which is structurally equivalent anyway?